src/Microsoft.OData.Core/UriParser/ODataQueryOptionParser.cs

```diff
@@ -506,21 +507,40 @@ private static SearchClause ParseSearchImplementation(string search, ODataUriPar
         }
 
         /// <summary>
-        /// Get query options according to case insensitive.
+        /// Gets query options according to case sensitivity and 
+        /// whether no dollar query options is enabled.
         /// </summary>
-        /// <param name="queryOptionName">The query option's name.</param>
-        /// <param name="value">The value for that query option.</param>
+        /// <param name="queryOptionName">The query option name.</param>
+        /// <param name="value">The value of the query option.</param>
         /// <returns>Whether value successfully retrived.</returns>
         private bool TryGetQueryOption(string queryOptionName, out string value)
         {
-            if (!this.Resolver.EnableCaseInsensitive)
+            value = null;
+            if (queryOptionName == null)
             {
-                return this.queryOptions.TryGetValue(queryOptionName, out value);
+                return false;
             }
 
-            value = null;
+            // Trim queryOptionName to prevent caller from passing in untrimmed name for comparison with 
+            // already trimmed keys in queryOptions dictionary.
+            string trimmedQueryOptionName = queryOptionName.Trim();
+
+            bool isCaseInsensitiveEnabled = this.Resolver.EnableCaseInsensitive;
+            bool isNoDollarQueryOptionsEnabled = this.Configuration.EnableNoDollarQueryOptions;
+
+            if (!isCaseInsensitiveEnabled && !isNoDollarQueryOptionsEnabled)
+            {
+                return this.queryOptions.TryGetValue(trimmedQueryOptionName, out value);
+            }
+
+            StringComparison stringComparison = isCaseInsensitiveEnabled ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
+
+            string noDollarQueryOptionName = (isNoDollarQueryOptionsEnabled && trimmedQueryOptionName.StartsWith(UriQueryConstants.DollarSign)) ?

```

Comment:
Code analysis will report an error for this name : "CSharp.Naming : The variable name 'noDollarQueryOptionName' begins with a prefix that looks like Hungarian notation. Remove the prefix or add it to the list of allowed prefixes."

So how about rename this to queryOptionNameWithoutDollar?